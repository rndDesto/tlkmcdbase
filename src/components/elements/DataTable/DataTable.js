import React from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scoped.css';

export default function DataTable(props) {
  const { className, column, data, } = props;
  const classes = [styles.root, className].filter(Boolean).join(' ');

  return (
    <table className={classes}>
      <thead>
        <tr>
          {column.map((item, idx) => <TableHeader item={item} key={idx} />)}
        </tr>
      </thead>
      <tbody>
        {data.map((item, idx) => <TableRow column={column} item={item} key={idx} />)}
      </tbody>
    </table>
  );
}

DataTable.defaultProps = {
  className: '',
  column: [],
  data: [],
};

DataTable.propTypes = {
  className: PropTypes.string,
  column: PropTypes.array,
  data: PropTypes.array,
};

export function TableHeader({ item }) {

  return (
    <th className={styles['table-header']}>{item.heading}</th>
  );
}

TableHeader.defaultProps = {
  item: {},
};

TableHeader.propTypes = {
  item: PropTypes.object,
};

export function TableRow({ column, item }) {
  return (
    <tr>
      {column.map((cItem, cIdx) => {
        const { value } = cItem;
        const newValue = typeof value === 'function' ? value(item) : item[value];

        return (
          <td key={cIdx}>
            {newValue || '-'}
          </td>
        );
      })}
    </tr>
  );
}

TableRow.defaultProps = {
  column: [],
  item: {},
};

TableRow.propTypes = {
  column: PropTypes.array,
  item: PropTypes.object,
};
