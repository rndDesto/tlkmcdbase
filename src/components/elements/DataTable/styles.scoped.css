.root {
  border-spacing: 0;
  text-align: left;
  width: 100%;
}

.root > thead > tr {
  background-color: var(--blue-color);
  text-align: left;
  border-radius: 0.1875rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}

.root th {
  padding-left: 2rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  text-align: left;
  color: var(--background-color-2);
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}

.root th:first-child {
  border-top-left-radius: 0.3125rem;
  border-bottom-left-radius: 0.3125rem;
  color: var(--background-color-2);
}

.root th:last-child {
  border-top-right-radius: 0.3125rem;
  border-bottom-right-radius: 0.3125rem;
}

.root > tbody > tr {
  background-color: #ffffff;
}

.root td {
  color: var(--text-color);
  font-size: 0.875rem;
  line-height: 1.125rem;
  padding: 0.75rem 0.5rem;
  max-width: 20rem;
  padding-left: 2rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  word-break: break-all;
}

.root > tbody > tr:nth-child(even) > td {
  background-color: rgba(232, 232, 232, 0.25);
  padding-left: 2rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}