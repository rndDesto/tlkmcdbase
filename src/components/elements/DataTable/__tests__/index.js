import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import DataTable, { TableHeader, TableRow } from '../DataTable';

const column = [
  { heading: 'Nama', value: 'name' },
  { heading: 'Email', value: 'email' },
  { heading: 'No. Handphone', value: 'phoneNumber' },
];

const data = [
  { name: 'Test', email: 'test@a.id', phoneNumber: '0811111111' },
  { name: 'Test', email: 'test@a.id', phoneNumber: '0811111111' },
];

describe('src/components/elements/DataTable', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<DataTable column={column} data={data} />);
    expect(tree).toMatchSnapshot();

    const result = DataTable({ ...DataTable.defaultProps, data: [], });
    expect(result.type).toBe('table');
  });

  test('TableHeader', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<TableHeader />);
    expect(tree).toMatchSnapshot();
  });

  test('TableRow', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<TableRow column={column} />);
    expect(tree).toMatchSnapshot();
  });
});
