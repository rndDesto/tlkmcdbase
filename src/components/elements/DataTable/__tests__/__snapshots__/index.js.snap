// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/DataTable TableHeader 1`] = `<th />`;

exports[`src/components/elements/DataTable TableRow 1`] = `
<tr>
  <td>
    -
  </td>
  <td>
    -
  </td>
  <td>
    -
  </td>
</tr>
`;

exports[`src/components/elements/DataTable render 1`] = `
<table
  className=""
>
  <thead>
    <tr>
      <TableHeader
        item={
          Object {
            "heading": "Nama",
            "value": "name",
          }
        }
      />
      <TableHeader
        item={
          Object {
            "heading": "Email",
            "value": "email",
          }
        }
      />
      <TableHeader
        item={
          Object {
            "heading": "No. Handphone",
            "value": "phoneNumber",
          }
        }
      />
    </tr>
  </thead>
  <tbody>
    <TableRow
      column={
        Array [
          Object {
            "heading": "Nama",
            "value": "name",
          },
          Object {
            "heading": "Email",
            "value": "email",
          },
          Object {
            "heading": "No. Handphone",
            "value": "phoneNumber",
          },
        ]
      }
      item={
        Object {
          "email": "test@a.id",
          "name": "Test",
          "phoneNumber": "0811111111",
        }
      }
    />
    <TableRow
      column={
        Array [
          Object {
            "heading": "Nama",
            "value": "name",
          },
          Object {
            "heading": "Email",
            "value": "email",
          },
          Object {
            "heading": "No. Handphone",
            "value": "phoneNumber",
          },
        ]
      }
      item={
        Object {
          "email": "test@a.id",
          "name": "Test",
          "phoneNumber": "0811111111",
        }
      }
    />
  </tbody>
</table>
`;
